init-container/init.c

```diff
@@ -0,0 +1,361 @@
+    while (size) {
+        ssize_t ret = read(fd, buf, size);
+        if (ret == 0) {
+            puts("Waiting for host connection ...");
+            sleep(1);
+            continue;
+        }
+        if (ret < 0) {
+            /* `errno` should be set. */
+            return -1;
+        }
+        buf = (char*)buf + ret;
+        size -= ret;
+    }
+    return 0;
+}
+
+static int recv_u64(int fd, uint64_t* res) {
+    return readn(fd, res, sizeof(*res));
+}
+
+static int recv_u32(int fd, uint32_t* res) {
+    return readn(fd, res, sizeof(*res));
+}
+
+static int recv_bytes(int fd, char** buf_ptr, uint64_t* size_ptr,
+                      bool is_cstring) {
+    uint64_t size = 0;
+
+    if (recv_u64(fd, &size) < 0) {
+        return -1;
+    }
+
+    char* buf = malloc(size + (is_cstring ? 1 : 0));
+    if (!buf) {
+        return -1;
+    }
+
+    if (readn(fd, buf, size) < 0) {
+        int tmp_errno = errno;
+        free(buf);
+        errno = tmp_errno;
+        return -1;
+    }
+
+    if (is_cstring) {
+        buf[size] = '\0';
+    }
+
+    *buf_ptr = buf;
+    *size_ptr = size;
+
+    return 0;
+}
+
+static void free_strings_array(char** array) {
+    if (array) {
+        for (size_t i = 0; array[i]; ++i) {
+            free(array[i]);
+        }
+        free(array);
+    }
+}
+
+static int recv_strings_array(int fd, char*** array_ptr) {
+    uint64_t size = 0;
+    int ret = -1;
+
+    if (recv_u64(fd, &size) < 0) {
+        return -1;
+    }
+
+    char** array = calloc(size + 1, sizeof(*array));
+    if (!array) {
+        return -1;
+    }
+
+    for (uint64_t i = 0; i < size; ++i) {
+        uint64_t tmp;
+        if (recv_bytes(fd, &array[i], &tmp, /*is_cstring=*/true) < 0) {
+            goto out;
+        }
+    }
+
+    *array_ptr = array;
+    array = NULL;
+    ret = 0;
+
+out:
+    if (array) {
+        int tmp_errno = errno;
+        free_strings_array(array);
+        errno = tmp_errno;
+    }
+    return ret;
+}
+
+static int writen(int fd, void* buf, size_t size) {
+    while (size) {
+        ssize_t ret = write(fd, buf, size);
+        if (ret == 0) {
+            puts("Waiting for host connection ...");
+            sleep(1);
+            continue;
+        }
+        if (ret < 0) {
+            /* `errno` should be set. */
+            return -1;
+        }
+        buf = (char*)buf + ret;
+        size -= ret;
+    }
+    return 0;
+}
+
+static void send_response(msg_id_t msg_id, uint32_t ret_val) {
+    struct msg_hdr resp = {
+        .msg_id = msg_id,
+        .type = !ret_val ? RESP_OK : RESP_ERR,
+    };
+
+    CHECK(writen(g_cmds_fd, &resp, sizeof(resp)));
+    if (ret_val) {
+        CHECK(writen(g_cmds_fd, &ret_val, sizeof(ret_val)));
+    }
+}
+
+static void print_strings_array(char** strings) {
+    for (unsigned i = 0; *strings; strings++, i++) {
+        printf("%u: %s\n", i, *strings);
+    }
+}
+
+static uint32_t spawn_new_process(uid_t uid, char* bin, char** argv,
```

Comment:
I guess it's still a WIP?

Reply:
This is just to show how to use the communication protocol and have a small test.